### PR TITLE
[Draft] Add module notification_rule

### PR DESCRIPTION
**Pull request by muehlings**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

There is no module for creating, editing or deleting notification rules.

## What is the new behavior?

[Draft] Module for creating, editing or deleting notification rules.

## Other information

Fourth try of generating Ansible modules from OpenAPI definition for Checkmk. This code is 99% generated from the checkmk.yaml file in SwaggerUI. Manually changed parts are around the [...] brackets. Module is based on Checkmk 2.2.0p20-CME YAML. Improvement in the class functions and better guessed variables/fields in my generator.
This endpoint has a BUG. It has absolutly no ID to lookup. Notification rules are always generated when the module is called! I opened a thread in the Checkmk forum.

### CODE REVIEW

**Comment by muehlings** [comment]
Playbook:
```
  - name: "Create notification_rule '{{item.id}}'"
    checkmk.general.notification_rule:
      "automation_secret": "{{automation_secret}}"
      "automation_user": "{{automation_user}}"
      "server_url": "{{server_url}}"
      "site": "{{site}}"
      "state": "{{item.state}}"

      "rule_config": "{{item.rule_config}}"
      "id": "{{item.id}}" # missing in endpoint schema, not idempotent atm

    loop: "{{ notification_rules|flatten(levels=1) }}"
    loop_control:
      label: "{{ item.id }}"
```

host_vars:
```
---
notification_rules:
- id: e12a6d61-08cc-42e2-a6b1-6edcdcb5158b
  rule_config:
    conditions:
      event_console_alerts:
        state: disabled
      match_check_types:
        state: disabled
      match_contact_groups:
        state: disabled
      match_exclude_hosts:
        state: disabled
      match_exclude_service_groups:
        state: disabled
      match_exclude_service_groups_regex:
        state: disabled
      match_exclude_services:
        state: disabled
      match_folder:
        state: disabled
      match_host_event_type:
        state: disabled
      match_host_groups:
        state: disabled
      match_host_labels:
        state: disabled
      match_host_tags:
        state: disabled
      match_hosts:
        state: disabled
      match_notification_comment:
        state: disabled
      match_only_during_time_period:
        state: disabled
      match_plugin_output:
        state: disabled
      match_service_event_type:
        state: disabled
      match_service_groups:
        state: disabled
      match_service_groups_regex:
        state: disabled
      match_service_labels:
        state: disabled
      match_service_levels:
        state: disabled
[...]
          graphs_per_notification:
            state: disabled
          info_to_be_displayed_in_the_email_body:
            state: disabled
          insert_html_section_between_body_and_table:
            state: disabled
          plugin_name: mail
          reply_to:
            state: disabled
          send_separate_notification_to_every_recipient:
            state: disabled
          sort_order_for_bulk_notificaions:
            state: disabled
          subject_for_host_notifications:
            state: disabled
          subject_for_service_notifications:
            state: disabled
          url_prefix_for_links_to_checkmk:
            state: disabled
    rule_properties:
      allow_users_to_deactivate:
        state: enabled
      comment: ''
      description: Notify all contacts of a host/service via HTML email (clone)
      do_not_apply_this_rule:
        state: enabled
      documentation_url: ''
  state: present
```

**Comment by alien999999999** [comment]
I have some questions on how to use this:

- how do you choose folder
- if you choose something else, like slack; how do you specify the other parameters like webhook?
- doc states that one should have a dedicated user for specific notifications, but how do i link this notification rule to a user "slack_notification" ?
- also, this seems to have some depths on these dicts, is there a way to use a simplified form?

Atm, i'm trying this and my task is like this:

```yaml
          - name: Add Slack notification
            checkmk.general.notification_rule:
                    rule_config:
                            location:
                                    folder: "/"
                    rule_properties:
                            description: "Slack notification"
                    notification_method:
                            notification_plugin:
                                    plugin_params:
                                            plugin_name: slack
                                            webhook_url: {{ checkmk.notifications.slack }}
                                            contact_users:
                                                    - notification_slack
                    server_url: "https://{{ checkmk.api_host }}/"
                    site: "{{ checkmk.site }}"
                    automation_user: "{{ checkmk.user }}"
                    automation_secret: "{{ checkmk_automation_secret }}"
                    validate_certs: false
            when: checkmk.notifications.slack is defined
```

File: plugins/modules/notification_rule.py [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+# i.e. the version is of the form "2.5.0" and not "2.4".
+version_added: "4.3.0"
+
+description:
+- Manage notification_rule in Checkmk.
+
+extends_documentation_fragment: [checkmk.general.common]
+
+options:
+    [...]
```
Comment (robin-checkmk) [inline plugins/modules/notification_rule.py:27]: Documentation is a must-have. Please align the documentation section with existing modules and the `options` section with the `module_args`.

File: plugins/modules/notification_rule.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+
+author:
+    - Stefan Mühling (@muehlings)
+"""
+
+EXAMPLES = r"""
+    [...]
```
Comment (robin-checkmk) [inline plugins/modules/notification_rule.py:34]: Please provide at least some basic examples.

**Review by robin-checkmk — CHANGES_REQUESTED** [review]

**Comment by alien999999999** [comment]
There is also nothing here on the **contact_selection** option, which apparently is required? I wonder if these defaults and parameters need to be changed in the notification module itself, instead of here.

**Comment by muehlings** [comment]
> I have some questions on how to use this:
> 
>     * how do you choose folder
As a matching option? See my example below. I tried to incorporate your parameters (folder/user/webhook) there.

> 
>     * if you choose something else, like slack; how do you specify the other parameters like webhook?
I describe a method how to design the required input later.


>     * doc states that one should have a dedicated user for specific notifications, but how do i link this notification rule to a user "slack_notification" ?
> 
>     * also, this seems to have some depths on these dicts, is there a way to use a simplified form?
I'm not a developer from Checkmk. Since the notifications allow a lot of parameters I don't think it's possible to simplify or "shortcut" any options here.

> Atm, i'm trying this and my task is like this:
> 
> ```yaml
>           - name: Add Slack notification
>             checkmk.general.notification_rule:
>                     rule_config:
>                             location:
>                                     folder: "/"
>                     rule_properties:
>                             description: "Slack notification"
>                     notification_method:
>                             notification_plugin:
>                                     plugin_params:
>                                             plugin_name: slack
>                                             webhook_url: {{ checkmk.notifications.slack }}
>                                             contact_users:
>                                                     - notification_slack
>                     server_url: "https://{{ checkmk.api_host }}/"
>                     site: "{{ checkmk.site }}"
>                     automation_user: "{{ checkmk.user }}"
>                     automation_secret: "{{ checkmk_automation_secret }}"
>                     validate_certs: false
>             when: checkmk.notifications.slack is defined
> ```
At first glance: It looks like an indention error. 

Thanks for trying my alpha state plugin. It is missing documentation and examples. 
The problem is: Each notification plugin is individual and needs its own rule_config dictionary. But I try to help you to achieve your desired result.

I set up my rules in a reference site. The details of the screenshot do not matter here. Everytime you enable a checkbox another dictionary or additional keys are necessary (look at "Match folder -> Main" later):
grafik

I then go to the SwaggerUI of that site and list all objects of that endpoint (/check_mk/api/1.0/domain-types/notification_rule/collections/all) and select the desired rule and copy it to a small python script to convert it to the according yaml notation. It is a bit difficult to select the object but after adding/deleting some brackets you should succeed:
```
{
        "rule_config": {
          "rule_properties": {
            "description": "Notify all contacts of Main folder via Slack",
            "comment": "",
            "documentation_url": "",
            "do_not_apply_this_rule": {
              "state": "disabled"
            },
            "allow_users_to_deactivate": {
              "state": "enabled"
            }
          },
          "notification_method": {
            "notify_plugin": {
              "option": "create_notification_with_the_following_parameters",
              "plugin_params": {
                "plugin_name": "slack",
                "disable_ssl_cert_verification": {
                  "state": "enabled"
                },
                "http_proxy": {
                  "state": "disabled"
                },
                "url_prefix_for_links_to_checkmk": {
                  "state": "enabled",
                  "value": {
                    "url": "http://mycheckmk-host/mysite/check_mk/",
                    "option": "manual"
                  }
                },
                "webhook_url": {
                  "option": "explicit",
                  "url": "https://mywebhook/api/vi"
                }
              }
            },
            "notification_bulking": {
              "state": "disabled"
            }
          },
          "contact_selection": {
            "all_contacts_of_the_notified_object": {
              "state": "enabled"
            },
            "all_users": {
              "state": "disabled"
            },
            "all_users_with_an_email_address": {
              "state": "disabled"
            },
            "the_following_users": {
              "state": "enabled",
              "value": [
                "notification_slack"
              ]
            },
            "members_of_contact_groups": {
              "state": "disabled"
            },
            "explicit_email_addresses": {
              "state": "disabled"
            },
            "restrict_by_custom_macros": {
              "state": "disabled"
            },
            "restrict_by_contact_groups": {
              "state": "disabled"
            }
          },
          "conditions": {
            "match_sites": {
              "state": "disabled"
            },
            "match_folder": {
              "state": "enabled",
              "value": "~"
            },
            "match_host_tags": {
              "state": "disabled"
            },
            "match_host_labels": {
              "state": "disabled"
            },
            "match_host_groups": {
              "state": "disabled"
            },
            "match_hosts": {
              "state": "disabled"
            },
            "match_exclude_hosts": {
              "state": "disabled"
            },
            "match_service_labels": {
              "state": "disabled"
            },
            "match_service_groups": {
              "state": "disabled"
            },
            "match_exclude_service_groups": {
              "state": "disabled"
            },
            "match_service_groups_regex": {
              "state": "disabled"
            },
            "match_exclude_service_groups_regex": {
              "state": "disabled"
            },
            "match_services": {
              "state": "disabled"
            },
            "match_exclude_services": {
              "state": "disabled"
            },
            "match_check_types": {
              "state": "disabled"
            },
            "match_plugin_output": {
              "state": "disabled"
            },
            "match_contact_groups": {
              "state": "disabled"
            },
            "match_service_levels": {
              "state": "disabled"
            },
            "match_only_during_time_period": {
              "state": "disabled"
            },
            "match_host_event_type": {
              "state": "disabled"
            },
            "match_service_event_type": {
              "state": "disabled"
            },
            "restrict_to_notification_numbers": {
              "state": "disabled"
            },
            "throttle_periodic_notifications": {
              "state": "disabled"
            },
            "match_notification_comment": {
              "state": "disabled"
            },
            "event_console_alerts": {
              "state": "disabled"
            }
          }
        }
      }
```
is then converted to
```
---
rule_configs:
  rule_config:
    conditions:
      event_console_alerts:
        state: disabled
      match_check_types:
        state: disabled
      match_contact_groups:
        state: disabled
      match_exclude_hosts:
        state: disabled
      match_exclude_service_groups:
        state: disabled
      match_exclude_service_groups_regex:
        state: disabled
      match_exclude_services:
        state: disabled
      match_folder:
        state: enabled
        value: '~'
      match_host_event_type:
        state: disabled
      match_host_groups:
        state: disabled
      match_host_labels:
        state: disabled
      match_host_tags:
        state: disabled
      match_hosts:
        state: disabled
      match_notification_comment:
        state: disabled
      match_only_during_time_period:
        state: disabled
      match_plugin_output:
        state: disabled
      match_service_event_type:
        state: disabled
      match_service_groups:
        state: disabled
      match_service_groups_regex:
        state: disabled
      match_service_labels:
        state: disabled
      match_service_levels:
        state: disabled
      match_services:
        state: disabled
      match_sites:
        state: disabled
      restrict_to_notification_numbers:
        state: disabled
      throttle_periodic_notifications:
        state: disabled
    contact_selection:
      all_contacts_of_the_notified_object:
        state: enabled
      all_users:
        state: disabled
      all_users_with_an_email_address:
        state: disabled
      explicit_email_addresses:
        state: disabled
      members_of_contact_groups:
        state: disabled
      restrict_by_contact_groups:
        state: disabled
      restrict_by_custom_macros:
        state: disabled
      the_following_users:
        state: enabled
        value:
        - notification_slack
    notification_method:
      notification_bulking:
        state: disabled
      notify_plugin:
        option: create_notification_with_the_following_parameters
        plugin_params:
          disable_ssl_cert_verification:
            state: enabled
          http_proxy:
            state: disabled
          plugin_name: slack
          url_prefix_for_links_to_checkmk:
            state: enabled
            value:
              option: manual
              url: http://mycheckmk-host/mysite/check_mk/
          webhook_url:
            option: explicit
            url: https://mywebhook/api/vi
    rule_properties:
      allow_users_to_deactivate:
        state: enabled
      comment: ''
      description: Notify all contacts of Main folder via Slack
      do_not_apply_this_rule:
        state: disabled
      documentation_url: ''
```

Full source with this example. It is not necessary to convert this string with json.dumps(). The yaml.dump() options at the end aligns the file to the known style:
```
#!/usr/bin/env python3

import yaml
import pprint

original_rule = {
        "rule_config": {
          "rule_properties": {
            "description": "Notify all contacts of Main folder via Slack",
            "comment": "",
            "documentation_url": "",
            "do_not_apply_this_rule": {
              "state": "disabled"
            },
            "allow_users_to_deactivate": {
              "state": "enabled"
            }
          },
          "notification_method": {
            "notify_plugin": {
              "option": "create_notification_with_the_following_parameters",
              "plugin_params": {
                "plugin_name": "slack",
                "disable_ssl_cert_verification": {
                  "state": "enabled"
                },
                "http_proxy": {
                  "state": "disabled"
                },
                "url_prefix_for_links_to_checkmk": {
                  "state": "enabled",
                  "value": {
                    "url": "http://mycheckmk-host/mysite/check_mk/",
                    "option": "manual"
                  }
                },
                "webhook_url": {
                  "option": "explicit",
                  "url": "https://mywebhook/api/vi"
                }
              }
            },
            "notification_bulking": {
              "state": "disabled"
            }
          },
          "contact_selection": {
            "all_contacts_of_the_notified_object": {
              "state": "enabled"
            },
            "all_users": {
              "state": "disabled"
            },
            "all_users_with_an_email_address": {
              "state": "disabled"
            },
            "the_following_users": {
              "state": "enabled",
              "value": [
                "notification_slack"
              ]
            },
            "members_of_contact_groups": {
              "state": "disabled"
            },
            "explicit_email_addresses": {
              "state": "disabled"
            },
            "restrict_by_custom_macros": {
              "state": "disabled"
            },
            "restrict_by_contact_groups": {
              "state": "disabled"
            }
          },
          "conditions": {
            "match_sites": {
              "state": "disabled"
            },
            "match_folder": {
              "state": "enabled",
              "value": "~"
            },
            "match_host_tags": {
              "state": "disabled"
            },
            "match_host_labels": {
              "state": "disabled"
            },
            "match_host_groups": {
              "state": "disabled"
            },
            "match_hosts": {
              "state": "disabled"
            },
            "match_exclude_hosts": {
              "state": "disabled"
            },
            "match_service_labels": {
              "state": "disabled"
            },
            "match_service_groups": {
              "state": "disabled"
            },
            "match_exclude_service_groups": {
              "state": "disabled"
            },
            "match_service_groups_regex": {
              "state": "disabled"
            },
            "match_exclude_service_groups_regex": {
              "state": "disabled"
            },
            "match_services": {
              "state": "disabled"
            },
            "match_exclude_services": {
              "state": "disabled"
            },
            "match_check_types": {
              "state": "disabled"
            },
            "match_plugin_output": {
              "state": "disabled"
            },
            "match_contact_groups": {
              "state": "disabled"
            },
            "match_service_levels": {
              "state": "disabled"
            },
            "match_only_during_time_period": {
              "state": "disabled"
            },
            "match_host_event_type": {
              "state": "disabled"
            },
            "match_service_event_type": {
              "state": "disabled"
            },
            "restrict_to_notification_numbers": {
              "state": "disabled"
            },
            "throttle_periodic_notifications": {
              "state": "disabled"
            },
            "match_notification_comment": {
              "state": "disabled"
            },
            "event_console_alerts": {
              "state": "disabled"
            }
          }
        }
      }

pprint.pprint(original_rule)

with open('rule_config.yaml', 'w') as yaml_file:
    print(yaml.dump({'rule_configs': original_rule}, default_flow_style=False, indent=2, allow_unicode=True, explicit_start=True), file=yaml_file)
print('Finished!')
```

The "id" is set dynamically on POST. Other endpoints allow getting and setting the id. This helps (at least me) to keep my Ansible and Checkmk relation idempotent. I set up 3 rules with 3 ids and can verify/modify/delete exact these 3 rules. I already reported the not setable notification_rule_id (aka id) in the Checkmk forum as a bug, but I have no answer yet. At the moment my Ansible tasks create all rules every time I execute it. 

I'm using PyYAML for the export:
```
pip3 install PyYAML
```

Kind regards and thanks for your feedback.

**Comment by alien999999999** [comment]
the others in the collection, seem to have something that state disabled options are not required to pass along; i don't know if this is done deeper in the code or in their ansible plugin, but that would make this a lot easier to work with

**Comment by muehlings** [comment]
Working example (state and id added at the end, required for my playbook):
```
/etc/ansible/host_vars/testsite/notification_rules.yaml
---
notification_rules:
- rule_config:
    conditions:
      event_console_alerts:
        state: disabled
      match_check_types:
        state: disabled
      match_contact_groups:
        state: disabled
      match_exclude_hosts:
        state: disabled
      match_exclude_service_groups:
        state: disabled
      match_exclude_service_groups_regex:
        state: disabled
      match_exclude_services:
        state: disabled
      match_folder:
        state: enabled
        value: '~'
      match_host_event_type:
        state: disabled
      match_host_groups:
        state: disabled
      match_host_labels:
        state: disabled
      match_host_tags:
        state: disabled
      match_hosts:
        state: disabled
      match_notification_comment:
        state: disabled
      match_only_during_time_period:
        state: disabled
      match_plugin_output:
        state: disabled
      match_service_event_type:
        state: disabled
      match_service_groups:
        state: disabled
      match_service_groups_regex:
        state: disabled
      match_service_labels:
        state: disabled
      match_service_levels:
        state: disabled
      match_services:
        state: disabled
      match_sites:
        state: disabled
      restrict_to_notification_numbers:
        state: disabled
      throttle_periodic_notifications:
        state: disabled
    contact_selection:
      all_contacts_of_the_notified_object:
        state: enabled
      all_users:
        state: disabled
      all_users_with_an_email_address:
        state: disabled
      explicit_email_addresses:
        state: disabled
      members_of_contact_groups:
        state: disabled
      restrict_by_contact_groups:
        state: disabled
      restrict_by_custom_macros:
        state: disabled
      the_following_users:
        state: enabled
        value:
        - notification_slack
    notification_method:
      notification_bulking:
        state: disabled
      notify_plugin:
        option: create_notification_with_the_following_parameters
        plugin_params:
          disable_ssl_cert_verification:
            state: enabled
          http_proxy:
            state: disabled
          plugin_name: slack
          url_prefix_for_links_to_checkmk:
            state: enabled
            value:
              option: manual
              url: http://mycheckmk-host/mysite/check_mk/
          webhook_url:
            option: explicit
            url: https://mywebhook/api/vi
    rule_properties:
      allow_users_to_deactivate:
        state: enabled
      comment: ''
      description: Notify all contacts of Main folder via Slack
      do_not_apply_this_rule:
        state: disabled
      documentation_url: ''
  state: present
  id: test_id
```

Playbook:
```
---
- name: 'Deployment of notification_rules to "{{appliance}}"'
  hosts: "{{appliance}}"
  gather_facts: False
  connection: local

#  # ====================== Prompt for variable "appliance" ======================
  vars_prompt:
  - name: appliance
    prompt: "Which appliance would you like to setup"
    private: no

  tasks:
  # ======================  (remote) ======================
  - name: "Create notification_rule '{{item.id}}'"
    checkmk.general.notification_rule:
      "automation_secret": "{{automation_secret}}"
      "automation_user": "{{automation_user}}"
      "server_url": "{{server_url}}"
      "site": "{{site}}"
      "state": "{{item.state}}"

      "rule_config": "{{item.rule_config}}"
      "id": "{{item.id}}"

    loop: "{{ notification_rules|flatten(levels=1) }}"
    loop_control:
      label: "{{ item.id }}"
```

Playbook result:
grafik

Checkmk result:
grafik

Clumsy like hell, but it works somehow. :)

The idempotency is not existing/working, but that is currently something I can't fix. My workaround is: 
```
- name: Remove current notifications and overwrite with my notification_rules with next task
  ansible.builtin.file:
    path: /omd/site/{{appliance}}/etc/check_mk/conf.d/wato/notifications.mk
    state: absent
```

**Comment by muehlings** [comment]
> the others in the collection, seem to have something that state disabled options are not required to pass along; i don't know if this is done deeper in the code or in their ansible plugin, but that would make this a lot easier to work with

I will try to leave out one key for a quick test. Then the initial rule_config can be a lot smaller.

**Comment by alien999999999** [comment]
So, the notification_bulking was not required, but if i didn't specify, i get ERROR 500 ; so I think it's something deeper in checkmk to be fixed.

in any case, i tested it, it worked for me; as you said it's not idempotent, but it's better than nothing; i was gonna try to check for it's existance, but i couldn't see a lookup version of this.

I feel like the **option: create_notification_with_the_following_parameters** is the culprit here; but i do recognize that most of this is a deeper issue inside check_mk

**Comment by alien999999999** [comment]
maybe semi-idempotency can be faked using notification_method; which would mean that you can't actually update it, but you can do state: present or absent to create or delete it?

**Comment by muehlings** [comment]
> So, the notification_bulking was not required, but if i didn't specify, i get ERROR 500 ; so I think it's something deeper in checkmk to be fixed.

> I feel like the **option: create_notification_with_the_following_parameters** is the culprit here; but i do recognize that most of this is a deeper issue inside check_mk

I'm currently investigating in ~/lib/check_mk/gui/plugins/openapi/endpoints/notification_rules/ and hope for some kind of ValueSpec somewhere else. 

This looks like the right source (for improving my code generation / OpenAPI-to-Ansible  converter):
```
from cmk.gui.plugins.wato.utils import notification_parameter_registry
```

**Comment by alien999999999** [comment]
So, I used your workaround to recreate the notifications.mk

but for the msteams plugin, it complains that the webhook_url.url is not an url type.

```
fatal: [checkmk]: FAILED! => {"changed": false, "msg": "400 - Bad Request: Parameter or validation failure. Details: b'{\"title\": \"Bad Request\", \"status\": 400, \"detail\": \"These fields have problems: rule_config\", \"fields\": {\"rule_config\": {\"notification_method\": {\"notify_plugin\": {\"plugin_params\": {\"webhook_url\": {\"url\": [\"Not a valid URL.\"]}}}}}}}'"}
```

While it's not a msteams webhook url, it is definately an url; so i don't know why it complains here? does this actually check for an exact msteams url?

**Comment by alien999999999** [comment]
btw: i do believe that robin-checkmk meant that you needed to add this documentation inside the code. though i don't know how you would document all the different notification plugins, tbh

**Comment by muehlings** [comment]
> btw: i do believe that robin-checkmk meant that you needed to add this documentation inside the code. though i don't know how you would document all the different notification plugins, tbh

That's the problem. :-}

**Comment by alien999999999** [comment]
well, good luck trying to get that doc in there and thanks for this code at least; i think it would be helpful if checkmk people would make some internal changes so that this would be shorter to configure.

At the very least, I am thankful for this code and I'll be using it :-)

**Comment by muehlings** [comment]
> So, I used your workaround to recreate the notifications.mk
> 
> but for the msteams plugin, it complains that the webhook_url.url is not an url type.
> 
> ```
> fatal: [checkmk]: FAILED! => {"changed": false, "msg": "400 - Bad Request: Parameter or validation failure. Details: b'{\"title\": \"Bad Request\", \"status\": 400, \"detail\": \"These fields have problems: rule_config\", \"fields\": {\"rule_config\": {\"notification_method\": {\"notify_plugin\": {\"plugin_params\": {\"webhook_url\": {\"url\": [\"Not a valid URL.\"]}}}}}}}'"}
> ```
> 
> While it's not a msteams webhook url, it is definately an url; so i don't know why it complains here? does this actually check for an exact msteams url?

I tested it with a working Webhook. This only works when set up directly in Checkmk. When I try to create it via SwaggerUI or Ansible the rule is broken:
grafik

These fields just disapeared:
grafik

**Comment by alien999999999** [comment]
you should also change the meta/run.yml from the collections and add notification_rule to the action_group , so that group defaults work for notification_rule too

**Comment by alien999999999** [comment]
this is really too bad that it could not be completed... i'm actually using this, it needs some more doc and changes in meta to be comitted i guess, but it's just closed, cause noone looked at it :-(

**Comment by robin-checkmk** [comment]
> this is really too bad that it could not be completed... i'm actually using this, it needs some more doc and changes in meta to be comitted i guess, but it's just closed, cause noone looked at it :-(

This has nothing to do with no one looking at it. The quality of this PR was still too low for a merge and the work needed to be done was beyond a quick fix. And most importantly did the author himself never finish the PR. We cannot simply take someone else's work and pretend it is ours. So as he decided not to move forward, there was nothing we could do.

**Comment by alien999999999** [comment]
I meant noone improved this patch, i know it's not good enough to submit, I commented on that. but, isn't this a modification of open source, which means, someone else can use this "open source" fragment and improve on it? even if it is unfinished? can a PR not have multiple authors?

**Comment by robin-checkmk** [comment]
> I meant noone improved this patch, i know it's not good enough to submit, I commented on that. but, isn't this a modification of open source, which means, someone else can use this "open source" fragment and improve on it? even if it is unfinished? can a PR not have multiple authors?

Ah, then misunderstood, thanks for clarifying!

You are right, anyone should be able to build on top of this PR. What I was thinking about primarily, is that I want to respect the authors work and give them the power to steer their own PR. 

Regardless, I am as sad as you are, if a PR goes stale and eventually closed, but having zombie PRs does no good either. So if you or anyone else can, I would love to see a new and healthy PR, that we can move forward together. 🙂

**Comment by muehlings** [comment]
I still have no time to do further work here. Feel free to fork my branch and finish it.

**Pull request closed without merging**